### PR TITLE
Grass curing tab

### PR DESCRIPTION
**Pull request by dgboss**
Adds the Grass Curing tab which displays the 'forecast' percent grass curing value and the value extracted from the Canadian Wildland Fire Information oSystem (CWFIS).

Fixes a minor where grass curing values would populate backwards in time when manually entering a new value.

No double-click support.
No Show/Hide columns support.
# Test Links:
[Landing Page](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3423.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 1]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
   }
 }
 
+export class GrassCuringFCWFISield implements ColDefGenerator {
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:84]:
```suggestion
export class GrassCuringCWFISField implements ColDefGenerator {
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 2]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
...
+
+  public generateColDef = () => {
+    return this.colDefBuilder.generateColDefWith(this.field, this.headerName, this.precision)
+  }
+
+  public generateColDefs = () => {
+    return [this.generateColDef()]
+  }
+
+  public static getInstance(): GrassCuringFCWFISield {
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:110]:
```suggestion
  public static getInstance(): GrassCuringCWFISField {
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 3]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
...
+    if (!GrassCuringFCWFISield.instance) {
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:111]:
```suggestion
    if (!GrassCuringCWFISField.instance) {
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 4]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
...
+      GrassCuringFCWFISield.instance = new GrassCuringFCWFISield()
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:112]:
```suggestion
      GrassCuringCWFISField.instance = new GrassCuringCWFISField()
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 5]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
...
+    }
+
+    return GrassCuringFCWFISield.instance
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:115]:
```suggestion
    return GrassCuringCWFISField.instance
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 6]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
   }
 }
 
+export class GrassCuringFCWFISield implements ColDefGenerator {
+  private static instance: GrassCuringFCWFISield
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:85]:
```suggestion
  private static instance: GrassCuringCWFISField
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 7]
```diff
@@ -160,3 +196,7 @@ export const MORECAST2_INDEX_FIELDS: ForecastColDefGenerator[] = [
   fwiField,
   dgrField
 ]
+
+export const MORECAST2_GRASS_CURING_FORCAST_FIELD: ForecastColDefGenerator = gcForecastField
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:200]:
```suggestion
export const MORECAST2_GRASS_CURING_FORECAST_FIELD: ForecastColDefGenerator = gcForecastField
```

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 8]
```diff
@@ -160,3 +196,7 @@ export const MORECAST2_INDEX_FIELDS: ForecastColDefGenerator[] = [
...
+
+export const MORECAST2_GRASS_CURING_CWFIS_FIELD: ColDefGenerator = GrassCuringFCWFISield.getInstance()
```
Comment (conbrad) [inline web/src/features/moreCast2/components/MoreCast2Column.tsx:202]:
```suggestion
export const MORECAST2_GRASS_CURING_CWFIS_FIELD: ColDefGenerator = GrassCuringCWFISField.getInstance()
```

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx [hunk 9]
```diff
@@ -397,7 +393,12 @@ const TabbedDataGrid = ({ morecast2Rows, fromTo, setFromTo }: TabbedDataGridProp
   // row has no actual value for the weather parameter of interest)
   const handleCellDoubleClick = (params: GridCellParams) => {
     const headerName = params.colDef.headerName as WeatherDeterminateType
-    if (!headerName || headerName === WeatherDeterminate.ACTUAL || headerName === WeatherDeterminate.FORECAST) {
+    if (
+      !headerName ||
+      headerName === WeatherDeterminate.ACTUAL ||
+      headerName === WeatherDeterminate.FORECAST ||
+      params.field.indexOf('grass') > -1
```
Comment (conbrad) [inline web/src/features/moreCast2/components/TabbedDataGrid.tsx:400]: What does an index of -1 represent here?

File: web/src/features/moreCast2/util.ts [hunk 10]
```diff
@@ -150,7 +150,12 @@ export const fillGrassCuring = (rows: MoreCast2Row[]): MoreCast2Row[] => {
   for (const row of rows) {
     const stationInfo = stationGrassMap.get(row.stationCode)
     // fill the grass curing forecast value, as long as it doesn't already have a value
-    if (stationInfo && row.grassCuringForecast && isNaN(row.grassCuringForecast.value)) {
+    if (
+      stationInfo &&
+      row.grassCuringForecast &&
+      isNaN(row.grassCuringForecast.value) &&
+      row.forDate > stationInfo.date
```
Comment (conbrad) [inline web/src/features/moreCast2/util.ts:157]: Is this the bug that was fixed? Is it easy to test?

File: web/src/features/moreCast2/components/MoreCast2Column.tsx (hunk 1 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/MoreCast2Column.tsx]: Oops, thanks!

File: web/src/features/moreCast2/util.ts (hunk 10 above)
Comment (dgboss) [reply web/src/features/moreCast2/util.ts:157]: Yes and yes! I've updated the tests to cover the fix. Thanks!

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx (hunk 9 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/TabbedDataGrid.tsx:400]: Sorry, I went old school on you there. string.indexOf will return -1 if the specified string is not included in `params.field`. The condition is analagous to the much more readable `params.field.includes('grass')`.

Comment (dgboss) [reply web/src/features/moreCast2/components/TabbedDataGrid.tsx]: Switch to use `.includes()`

**Review by brettedw — APPROVED** [review]
Looks good!

**Review by conbrad — APPROVED** [review]

File: web/src/features/moreCast2/components/MoreCast2Column.tsx [hunk 11]
```diff
@@ -81,6 +81,41 @@ export class DateForecastField implements ColDefGenerator {
...
+
+  public generateColDef = () => {
+    return this.colDefBuilder.generateColDefWith(this.field, this.headerName, this.precision)
+  }
+
+  public generateColDefs = () => {
+    return [this.generateColDef()]
+  }
+
+  public static getInstance(): GrassCuringCWFISield {
```
[the remark this answers is not in the document]
Comment (dgboss) [reply web/src/features/moreCast2/components/MoreCast2Column.tsx:110]: Gah, still! Fix incoming.

**Pull request merged**